In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [2]:
folder=[]
def img_read(number,loop):
    for i in range(loop):
        a=plt.imread("C:\\Users\\STACBD\\bangla-money\\Training\\"+str(number)+"\\"+str(number)+"_"+str(i)+".jpg")
        a=np.array(a)
        folder.append(a)
    ax=np.array(folder)
    return a 

In [14]:
folder=np.array(folder)
folder.shape

(1664, 120, 250, 3)

We can see that our image is a tensor of rank 4, or we could say a 4 dimensional array with dimensions 1664 x 120 x 250 x 3 which correspond to the batch size, height, width and channels respectively.

In [3]:
y=[]
def label(number,loop):  
    for i in range(loop):
        if number==1:
            y.append(1)
        elif number==2:
            y.append(2)
        elif number==5:
            y.append(5)
        elif number==10:
            y.append(10)
        elif number==20:
            y.append(20)
        elif number==50:
            y.append(50)
        elif number==100:
            y.append(100)
        elif number==500:
            y.append(500)
        elif number==1000:
            y.append(1000)
    return y 

In [4]:
one_taka=img_read(1,100)
one=label(1,100)

In [5]:
two_taka=img_read(2,212)
two=label(2,212)

In [6]:
five_taka=img_read(5,212)
five=label(5,212)

In [7]:
ten_taka=img_read(10,212)
ten=label(10,212)

In [8]:
twenty_taka=img_read(20,212)
twenty=label(20,212)

In [9]:
fifty_taka=img_read(50,212)
fifty=label(50,212)

In [10]:
hundred_taka=img_read(100,203)
hundred=label(100,203)

In [11]:
five_hundred_taka=img_read(500,135)
five_hundred=label(500,135)

In [12]:
one_thousand_taka=img_read(1000,166)
one_thousand=label(1000,166)

In [13]:
y=np.array(y)
print(y)
y.shape

[   1    1    1 ... 1000 1000 1000]


(1664,)

In [15]:
X=np.array(folder)

In [16]:
def rgb2gray(rgb):
    return np.dot(rgb[...,:3], [0.299, 0.587, 0.144])

This above code converts an image with RGB channels into an image with a single grayscale channel.

In [17]:
X=rgb2gray(folder)

In [18]:
X.shape

(1664, 120, 250)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.2)

print("X_train: "+str(X_train.shape))
print("X_test: "+str(X_test.shape))

print("y_train: "+str(y_train.shape))
print("y_test: "+str(y_test.shape))

In [22]:
from random import randrange

In [30]:
# Split a dataset into a train and test set
def train_test_split(dataset, split=0.2):
    train = list()
    train_size = split * len(dataset)
    dataset_copy = list(dataset)
    while len(train) < train_size:
        index = randrange(len(dataset_copy))
        train.append(dataset_copy.pop(index))
    return train, dataset_copy


y_test, y_train = train_test_split(y)
y_test=np.array(y_test)
y_train=np.array(y_train)
print("y_train:   "+str(len(y_train)))
print("y_test:    "+str(len(y_test)))

print("\n")

X_test, X_train = train_test_split(X)
X_test=np.array(X_test)
X_train=np.array(X_train)
print("X_train:   "+str(len(X_train)))
print("X_test:    "+str(len(X_test)))

y_train:   1331
y_test:    333


X_train:   1331
X_test:    333


In [31]:
print("X_train: "+str(X_train.shape))
print("X_test: "+str(X_test.shape))

print("y_train: "+str(y_train.shape))
print("y_test: "+str(y_test.shape))

X_train: (1331, 120, 250)
X_test: (333, 120, 250)
y_train: (1331,)
y_test: (333,)


In [32]:
X_train = np.reshape(X_train, (X_train.shape[0], -1))
X_test = np.reshape(X_test, (X_test.shape[0], -1))

In [33]:
print("X_train: "+str(X_train.shape))
print("X_test: "+str(X_test.shape))

X_train: (1331, 30000)
X_test: (333, 30000)


In [34]:
from collections import Counter

In [35]:
def euclidean_distance(x1, x2):
        return np.sqrt(np.sum((x1 - x2)**2))


class KNN:

    def __init__(self, k=3):
        self.k = k

    def fit(self, X, y):
        self.X_train = X
        self.y_train = y

    def predict(self, X):
        y_pred = [self._predict(x) for x in X]
        return np.array(y_pred)

    def _predict(self, x):
        # Compute distances between x and all examples in the training set
        distances = [euclidean_distance(x, x_train) for x_train in self.X_train]
        # Sort by distance and return indices of the first k neighbors
        k_idx = np.argsort(distances)[:self.k]
        # Extract the labels of the k nearest neighbor training samples
        k_neighbor_labels = [self.y_train[i] for i in k_idx]  
        # return the most common class label
        most_common = Counter(k_neighbor_labels).most_common(1)
        return most_common[0][0]

In [36]:
def accuracy(y_test, y_pred):
    accuracy = np.sum(y_test == y_pred) / len(y_test)*100.0
    return accuracy

In [37]:
k = 5
clf = KNN(k=k)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print("custom KNN classification accuracy", accuracy(y_test, y_pred))

custom KNN classification accuracy 10.21021021021021


if you choose an even k, there is a risk of a tie in the decision of which class you should set a new instance. This is why the k is usually odd - no ties.

In [28]:
k_evaluations = []

for k in range(1, 22, 2):
    clf = KNN(k=k)
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    
        
    curr_accuracy = accuracy(y_test, y_pred)
    k_evaluations.append((k, curr_accuracy))
    
k_evaluations

[(1, 75.07507507507508),
 (3, 68.76876876876878),
 (5, 65.16516516516516),
 (7, 63.36336336336337),
 (9, 61.261261261261254),
 (11, 58.25825825825825),
 (13, 57.35735735735735),
 (15, 58.55855855855856),
 (17, 55.85585585585585),
 (19, 53.453453453453456),
 (21, 53.753753753753756)]

Best KNN decide 